File: 백준/Gold/1956. 운동/운동.cc
#include <vector>
#include <queue>
#include <string.h>
#include <algorithm>
#include <iostream>
#include <climits>
#include <map>
#include <stack>
#include <math.h>

using namespace std;
#define FASTIO ios::sync_with_stdio(false); cin.tie(0); cout.tie(0)
#define LL long long
#define ULL unsigned long long

// const int INF = 1987654321;
const int INF = 1e9;
const int MOD = 1000000007;
const int MAX_V = 401;

int n, m;
int adj[MAX_V][MAX_V];

void floyd() {
    for(int k = 1; k <= n; k++) {
        for(int i = 1; i <= n; i++) {
            for(int j = 1; j <= n; j++) {
                adj[i][j] = min(adj[i][j], adj[i][k] + adj[k][j]);
            }
        }
    }
}

int main(void) {
    FASTIO;

    cin >> n >> m;
    for(int i = 0; i <= n; i++) {
        fill(adj[i], adj[i] + n + 1, INF);
        adj[i][i] = 0;
    }
    while(m--) {
        int a, b, c;
        cin >> a >> b >> c;
        adj[a][b] = c;
    }
    floyd();

    int res = INF;
    for(int i = 1; i <= n; i++) {
        for(int j = 1; j < i; j++) {
            res = min(res, adj[i][j] + adj[j][i]);
        }
    }
    if(res >= INF)
        cout << "-1\n";
    else
        cout << res << '\n';

    return 0;
}

File: 백준/Gold/1956. 운동/운동.java
import java.util.*;
import java.util.stream.Collector;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.io.*;

public class Main {

    static final int INF = (int)1e9;
    static int v, e;
    static int[][] adj;


    static void floyd() {
        for(int k = 0; k < v; k++) {
            for(int i = 0; i < v; i++) {
                for(int j = 0; j < v; j++) {
                    adj[i][j] = Math.min(adj[i][j], adj[i][k] + adj[k][j]);
                }
            }
        }
    }

    static int calcMin() {
        int ret = INF;
        for(int i = 0; i < v; i++) {
            for(int j = i + 1; j < v; j++) {
                ret = Math.min(ret, adj[i][j] + adj[j][i]);
            }
        }
        return ret;
    }
    
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        String[] input = br.readLine().split(" ");
        v = Integer.parseInt(input[0]);
        e = Integer.parseInt(input[1]);

        adj = new int[v][v];
        for(int i = 0; i < v; i++) {
            Arrays.fill(adj[i], INF);
            adj[i][i] = 0;
        }

        for(int i = 0; i < e; i++) {
            input = br.readLine().split(" ");
            int src = Integer.parseInt(input[0]);
            int dst = Integer.parseInt(input[1]);
            int cost = Integer.parseInt(input[2]);
            src--; dst--;
            adj[src][dst] = cost;
        }
        floyd();
        int res = calcMin();
        if(res == INF)
            bw.write("-1\n");
        else
            bw.write(res + "\n");        

        bw.flush();
        bw.close();
    }
}
